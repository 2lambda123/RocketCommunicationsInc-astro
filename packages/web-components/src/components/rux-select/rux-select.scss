@use '../../common/functional-components/FormFieldMessage/form-field-message.scss';

:host {
    display: block;
    scrollbar-color: var(--color-border-interactive-muted)
        var(--color-background-surface-default);
    ::-webkit-scrollbar {
        width: var(--spacing-4); //16px
        height: var(--spacing-4); //16px
        background-color: transparent;
    }

    ::-webkit-scrollbar-thumb {
        background-color: var(
            --color-border-interactive-muted,
            rgb(43, 101, 155)
        );
        border-radius: var(--progress-radius-inner);
        border: 3px solid transparent;
        background-clip: padding-box;
    }

    /* visually "centers" because the dark edge of the shadow gives the illusion this is offset */
    ::-webkit-scrollbar-thumb:vertical {
        border-left-width: var(--border-width-lg); //4px
    }

    ::-webkit-scrollbar-thumb:horizontal {
        border-top-width: var(--border-width-lg); //4px
    }

    ::-webkit-scrollbar-thumb:active,
    ::-webkit-scrollbar-thumb:hover {
        background-color: var(
            --color-background-interactive-default,
            rgb(58, 129, 191)
        );
    }

    ::-webkit-scrollbar-track,
    ::-webkit-scrollbar-corner {
        background-color: var(
            --color-background-surface-default,
            rgb(27, 45, 62)
        );

        box-shadow: var(--scrollbar-shadow-inner-vertical);
    }

    ::-webkit-scrollbar-track:vertical {
        box-shadow: var(--scrollbar-shadow-inner-vertical);
    }

    ::-webkit-scrollbar-track:horizontal {
        box-shadow: var(--scrollbar-shadow-inner-horizontal);
    }
}

.hidden,
:host([hidden]) {
    display: none;
}

label {
    display: inline-block;
    margin-bottom: calc(var(--spacing-3) - 2px); //10px
    color: var(--color-text-primary);
    font-family: var(--font-control-body-1-font-family);
    font-size: var(--font-control-body-1-font-size);
    font-weight: var(--font-control-body-1-font-weight);
    letter-spacing: var(--font-control-body-1-letter-spacing);
    .rux-label__asterisk {
        margin-left: var(--spacing-1); //4px
    }
}

.rux-select:not(.rux-select--multiple) {
    background-image: url("data:image/svg+xml,%3Csvg width='31' height='30' viewBox='0 0 31 30' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M11.6875 14.6375L14.925 17.875C15.4125 18.3625 16.2 18.3625 16.6875 17.875L19.925 14.6375C20.7125 13.85 20.15 12.5 19.0375 12.5H12.5625C11.45 12.5 10.9 13.85 11.6875 14.6375Z' fill='%234DACFF'/%3E%3C/svg%3E%0A"),
        linear-gradient(
            to left,
            var(--color-background-surface-selected) var(--spacing-8),
            var(--color-background-base-default) var(--spacing-8)
        );
    // background-image: url('data:image/svg+xml,%3Csvg%20width%3D%2210%22%20height%3D%225%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%3Cpath%20fill%3D%22%234dacff%22%20d%3D%22M0%200h10L5%205z%22%20fill-rule%3D%22evenodd%22%2F%3E%3C%2Fsvg%3E'),
    //     linear-gradient(
    //         to left,
    //         var(--color-background-surface-selected) var(--spacing-8),
    //         var(--color-background-base-default) var(--spacing-8)
    //     );
    background-position: center right var(--spacing-025),
        center left var(--spacing-0);
    background-repeat: no-repeat;
    &:active:not(:disabled) {
        background-image: url("data:image/svg+xml,%3Csvg width='31' height='30' viewBox='0 0 31 30' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M11.6875 14.6375L14.925 17.875C15.4125 18.3625 16.2 18.3625 16.6875 17.875L19.925 14.6375C20.7125 13.85 20.15 12.5 19.0375 12.5H12.5625C11.45 12.5 10.9 13.85 11.6875 14.6375Z' fill='%234DACFF'/%3E%3C/svg%3E%0A"),
            linear-gradient(
                to left,
                var(--color-background-surface-selected) var(--spacing-8),
                var(--color-background-base-default) var(--spacing-8)
            );
    }
    &:hover {
        cursor: pointer;
    }
}

.rux-select {
    box-sizing: border-box;
    position: relative;
    background: var(--color-background-base-default);
    appearance: none;
    -webkit-appearance: none;
    -moz-appearance: none;
    width: 100%;
    box-shadow: var(--color-border-interactive-muted) 0 0 0 1px inset; //replaces border so the border width does not count towards overall component height
    border: none;
    border-radius: var(--radius-base);
    color: var(--color-background-interactive-default);
    font-family: var(--font-control-body-1-font-family);
    font-size: var(--font-control-body-1-font-size);
    font-weight: var(--font-control-body-1-font-weight);
    line-height: var(--font-control-body-1-line-height);
    letter-spacing: var(--font-control-body-1-letter-spacing);
    user-select: none;

    &--small {
        padding: var(--spacing-1) var(--spacing-2); //4px 8px
    }
    &--medium {
        padding: var(--spacing-2); //8px
    }
    &--large {
        padding: var(--spacing-3) var(--spacing-2); //12px 8px
    }
    &--multiple {
        background: var(--color-background-base-default);
        padding: 0;
        option {
            padding: calc(var(--spacing-2) - 1px) var(--spacing-0)
                calc(var(--spacing-2) - 1px) var(--spacing-4); // 7px 0px 7px 16px
            text-indent: var(--spacing-4);
            &:hover {
                color: var(--color-background-interactive-hover);
                background-color: var(--color-background-surface-hover);
                cursor: pointer;
            }
            &:checked {
                background: var(--color-background-surface-selected);
                color: var(--color-text-primary);
            }
            &:disabled {
                cursor: not-allowed;
                &:hover {
                    color: var(--color-background-interactive-default);
                    background-color: linear-gradient(
                        to left,
                        var(--color-background-surface-selected)
                            var(--spacing-8),
                        var(--color-background-base-default) var(--spacing-8)
                    );
                }
            }
        }
    }
    &--invalid {
        box-shadow: var(--color-text-error) 0 0 0 1px inset; //replaces border so the border width does not count towards overall component height
        border: none;
    }

    &:hover {
        box-shadow: var(--color-background-interactive-hover) 0 0 0 1px inset; //replaces border so the border width does not count towards overall component height
        border: none;
    }

    &:focus {
        outline: none;
        box-shadow: var(--color-background-interactive-hover) 0 0 0 1px inset; //replaces border so the border width does not count towards overall component height
        border: none;
    }

    &:disabled {
        opacity: 0.4;
        cursor: not-allowed;
        &:hover {
            box-shadow: var(--color-border-interactive-muted) 0 0 0 1px inset; //replaces border so the border width does not count towards overall component height
            border: none;
        }
    }

    &::-ms-expand {
        display: none;
    }

    &:focus::-ms-value {
        background: transparent;
    }

    option {
        opacity: 1;
        color: var(--color-background-interactive-default);
        border-radius: 0;

        &:hover {
            color: var(--color-background-interactive-hover);
            background-color: var(--select-menu-option-hover-background-color);
        }

        &:focus {
            color: var(--color-text-primary);
            background-color: var(--color-background-surface-selected);
        }
        &:disabled {
            cursor: not-allowed;
            //var(--select-menu-text-color) with 0.4 opacity
            color: #4dacff66;
        }
    }

    optgroup {
        text-indent: var(--spacing-4);
        color: var(--select-menu-text-color);
        font-family: var(--font-body-1-font-family);
        font-size: var(--font-body-1-font-size);
        font-weight: var(--font-body-1-font-weight);
        letter-spacing: var(--font-body-1-letter-spacing);
        font-style: normal;
    }
}
