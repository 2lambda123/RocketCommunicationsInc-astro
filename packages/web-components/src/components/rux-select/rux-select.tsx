/* eslint react/jsx-no-bind: 0 */ // --> OFF
import {
    Component,
    Element,
    Host,
    h,
    Prop,
    Event,
    EventEmitter,
    Watch,
    Listen,
    State,
    Method,
} from '@stencil/core'
import { FormFieldInterface } from '../../common/interfaces.module'
import { hasSlot, renderHiddenSelect } from '../../utils/utils'

/**
 * @slot (default) - The select options
 * @slot label - The select label
 * @slot help-text -  the help text
 * @slot error-text -  the error text
 * @part error-text - The error text element
 * @part form-field - The form-field wrapper container
 * @part help-text - The help text element
 * @part label - The select label
 * @part select - The select element
 * @part required - The asterisk when required is true
 */
@Component({
    tag: 'rux-select',
    styleUrl: 'rux-select.scss',
    shadow: true,
})
export class RuxSelect implements FormFieldInterface {
    private slotContainer?: HTMLElement
    private selectEl!: HTMLSelectElement

    @Element() el!: HTMLRuxSelectElement
    @State() hasLabelSlot = false
    @State() hasHelpSlot = false
    @State() hasErrorSlot = false

    /**
     * Disables the select menu via HTML disabled attribute. Select menu takes on a distinct visual state. Cursor uses the not-allowed system replacement and all keyboard and mouse events are ignored.
     */
    @Prop({ reflect: true }) disabled: boolean = false

    /**
     * Sets the field as required
     */
    @Prop({ reflect: true }) required: boolean = false

    /**
     * The select label text. For HTML content, use the `label` slot instead.
     */
    @Prop() label?: string

    /**
     * Id for the Select Input
     */
    @Prop({ attribute: 'input-id' }) inputId?: string

    /**
     * Id for the Label
     */
    @Prop({ attribute: 'label-id' }) labelId?: string

    /**
     * Presentational only. Renders the Select Menu as invalid.
     */
    @Prop({ reflect: true }) invalid: boolean = false

    /**
     * Enables multiselect
     */
    @Prop({ reflect: true }) multiple: boolean = false

    /**
     * Sets the Name of the Input Element
     */
    @Prop({ reflect: true }) name = ''

    /**
     * The value of the selected option. If multiple is true, this is an array.
     */
    @Prop({ mutable: true }) value?: string | string[]

    /**
     * The help or explanation text
     */
    @Prop({ attribute: 'help-text' }) helpText?: string

    /**
     * The validation error text
     */
    @Prop({ attribute: 'error-text' }) errorText?: string

    /**
     * The size of rux-select
     */
    @Prop({ reflect: true }) size?: 'small' | 'medium' | 'large' = 'medium'

    /**
     * Styles the select menu as a inline, borderless variant.
     */
    @Prop({ reflect: true }) inline?: boolean = false

    /**
     * Event Emitted when the Value of the Select is Changed
     */
    @Event({ eventName: 'ruxchange' })
    ruxSelectChanged!: EventEmitter<void>

    /**
     * Fired when an element has lost focus - [HTMLElement/blur_event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)
     */
    @Event({ eventName: 'ruxblur' }) ruxBlur!: EventEmitter

    @Watch('value')
    onValueChange() {
        this._syncOptionsFromValue()
    }

    @Watch('label')
    handleLabelChange() {
        this._handleLabelSlotChange()
    }

    @Listen('rux-option-group-changed')
    handleGroupChange() {
        this._syncOptionsToNativeSelect()
        this._syncOptionsFromValue()
    }

    @Listen('rux-option-changed')
    handleOptionChange() {
        this._syncOptionsToNativeSelect()
    }

    /**
     * Sets element as focused
     */
    @Method()
    async setFocus(options?: FocusOptions) {
        this.selectEl.focus(options)
    }

    connectedCallback() {
        this._handleSlotChange = this._handleSlotChange.bind(this)
        this._handleLabelSlotChange = this._handleLabelSlotChange.bind(this)
    }

    componentWillLoad() {
        this._handleLabelSlotChange()
        if (this.value) {
            this._handleSlotChange()
        }
    }

    get hasLabel() {
        return this.label ? true : this.hasLabelSlot
    }

    private _onBlur = () => {
        this.ruxBlur.emit()
    }

    private _handleLabelSlotChange() {
        this.hasLabelSlot = hasSlot(this.el, 'label')
    }

    private async _handleSlotChange() {
        await this._syncOptionsToNativeSelect()
        await this._syncOptionsFromValue()
        this.hasErrorSlot = hasSlot(this.el, 'error-text')
        this.hasHelpSlot = hasSlot(this.el, 'help-text')
    }

    /**
     * The native select element doesn't play nicely with slots. If an <option> isn't a direct child element, it won't render properly.
     * As a solution, we expose a slot outside the shadow-DOMed <select> and then manually copy the contents inside the shadow DOM.
     *
     * A RuxOptionGroup component is required because onSlotchange won't fire if we use the native <optgroup> and we change just its options.
     * RuxOptionGroup exists only to fire a change event that we can listen to.
     */
    private _syncOptionsToNativeSelect() {
        const slot = this.slotContainer?.querySelector(
            'slot'
        ) as HTMLSlotElement

        if (slot) {
            this.selectEl.innerHTML = ''

            const assignedElements = slot.assignedElements({
                flatten: true,
            }) as HTMLElement[]

            assignedElements.map((item: any) => {
                const option = item
                if (option.tagName.toLowerCase() === 'rux-option') {
                    this._appendOptionToNativeSelect(
                        option.label,
                        option.value,
                        option.disabled,
                        this.selectEl,
                        option.title
                    )
                }

                if (option.tagName.toLowerCase() === 'rux-option-group') {
                    const children = [
                        ...Array.from(option.querySelectorAll('rux-option')),
                    ] as HTMLRuxOptionElement[]
                    this._appendOptGroupToNativeSelect(
                        option.label ? option.label : 'Group',
                        children,
                        option.title
                    )
                }
            })
        }
        return Promise.resolve()
    }

    private _appendOptGroupToNativeSelect(
        groupName: string,
        children: HTMLRuxOptionElement[],
        title?: string
    ) {
        const group = Object.assign(document.createElement('optgroup'), {
            label: groupName,
            title: title,
        })

        children.map((option: any) => {
            this._appendOptionToNativeSelect(
                option.label,
                option.value,
                option.disabled,
                group,
                option.title
            )
            this.selectEl.appendChild(group)
        })

        this.selectEl.appendChild(group)
    }

    private _appendOptionToNativeSelect(
        label: string,
        value: string,
        disabled: boolean,
        target: HTMLSelectElement | HTMLOptGroupElement,
        title?: string
    ) {
        const item = Object.assign(document.createElement('option'), {
            innerHTML: label ? label : '',
            value: value,
            disabled: disabled,
            title: title,
        })
        target.appendChild(item)
    }

    private _syncOptionsFromValue() {
        if (this.selectEl) {
            const options = [
                ...Array.from(this.selectEl.querySelectorAll('option')),
            ]
            options.map((option: HTMLOptionElement) => {
                if (Array.isArray(this.value)) {
                    option.selected = this.value.includes(option.value)
                } else {
                    option.selected = option.value === this.value
                }
            })
        }
        return Promise.resolve()
    }

    private _onChange(e: Event) {
        const target = e.target as HTMLSelectElement

        const values = [...target.options]
            .filter((option) => {
                return option.selected
            })
            .map((option) => {
                return option.value
            })

        if (values.length === 1) {
            this.value = values[0]
        } else {
            this.value = values
        }

        this.ruxSelectChanged.emit()
    }

    render() {
        const {
            disabled,
            required,
            label,
            inputId,
            labelId,
            invalid,
            name,
            multiple,
            errorText,
            helpText,
            hasHelpSlot,
            hasErrorSlot,
            _handleSlotChange,
        } = this
        renderHiddenSelect(true, this.el, this.name, this.value, this.disabled)
        return (
            <Host>
                <label
                    id={labelId}
                    htmlFor={inputId}
                    aria-hidden={this.hasLabel ? 'false' : 'true'}
                    part="label"
                >
                    <span class={{ hidden: !this.hasLabel }}>
                        <slot
                            onSlotchange={this._handleLabelSlotChange}
                            name="label"
                        >
                            {label}
                            {this.required && (
                                <span
                                    part="required"
                                    class="rux-label__asterisk"
                                >
                                    &#42;
                                </span>
                            )}
                        </slot>
                    </span>
                </label>
                <select
                    class={{
                        'rux-select': true,
                        'rux-select--small': this.size === 'small',
                        'rux-select--medium': this.size === 'medium',
                        'rux-select--large': this.size === 'large',
                        'rux-select--invalid': invalid,
                        'rux-select--multiple': multiple,
                        'rux-select--inline': this.inline === true,
                    }}
                    ref={(el) => (this.selectEl = el as HTMLSelectElement)}
                    id={inputId}
                    disabled={disabled}
                    required={required}
                    multiple={multiple}
                    name={name}
                    onChange={(e) => this._onChange(e)}
                    onBlur={this._onBlur}
                    part="select"
                ></select>
                <div
                    aria-hidden="true"
                    class="hidden"
                    ref={(el) => (this.slotContainer = el)}
                >
                    <slot onSlotchange={this._handleSlotChange}></slot>
                </div>
                <div
                    class={{
                        'rux-error-text': !!errorText || hasErrorSlot,
                        hidden: !errorText && !hasErrorSlot,
                    }}
                    part="error-text"
                >
                    <svg
                        fill="none"
                        width="14"
                        height="14"
                        xmlns="http://www.w3.org/2000/svg"
                        viewBox="0 0 14 14"
                    >
                        <path
                            fill-rule="evenodd"
                            clip-rule="evenodd"
                            d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z"
                            fill="currentColor"
                        />
                    </svg>
                    <slot name="error-text" onSlotchange={_handleSlotChange}>
                        {errorText}
                    </slot>
                </div>
                <div
                    class={{
                        'rux-help-text':
                            (!!helpText || hasHelpSlot) &&
                            (!errorText || !hasErrorSlot),
                        hidden:
                            (!helpText && !hasHelpSlot) ||
                            !!errorText ||
                            hasErrorSlot,
                    }}
                    part="help-text"
                >
                    <slot name="help-text" onSlotchange={_handleSlotChange}>
                        {helpText}
                    </slot>
                </div>
            </Host>
        )
    }
}
