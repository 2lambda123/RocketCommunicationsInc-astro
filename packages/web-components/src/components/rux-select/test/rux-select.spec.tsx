import { newSpecPage } from '@stencil/core/testing'
import { RuxSelect } from '../rux-select'

describe('rux-select', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxSelect],
            html: `
            <rux-select label="test">
                <rux-option label="one"></rux-option>
                <rux-option label="two"></rux-option>
            </rux-select>
      `,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('renders option groups', async () => {
        const page = await newSpecPage({
            components: [RuxSelect],
            html: `
            <rux-select>
              <rux-option label="outside option"></rux-option>
              <rux-option-group label="Group one">
                <rux-option label="inside option"></rux-option>
              </rux-option-group>
              <rux-option label="outside option"></rux-option>
            </rux-select>
          `,
        })
        expect(page.root).toMatchSnapshot()
    })
})
