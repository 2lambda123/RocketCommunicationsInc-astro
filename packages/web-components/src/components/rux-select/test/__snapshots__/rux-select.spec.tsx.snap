// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-select renders 1`] = `
<rux-select label="test" size="medium">
  <mock:shadow-root>
    <label aria-hidden="false" part="label">
      <span>
        <slot name="label">
          test
        </slot>
      </span>
    </label>
    <select class="rux-select rux-select--medium" part="select"></select>
    <div aria-hidden="true" class="hidden">
      <slot></slot>
    </div>
    <div class="hidden" part="error-text">
      <svg fill="none" height="14" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
        <path clip-rule="evenodd" d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z" fill="currentColor" fill-rule="evenodd"></path>
      </svg>
      <slot name="error-text"></slot>
    </div>
    <div class="hidden" part="help-text">
      <slot name="help-text"></slot>
    </div>
  </mock:shadow-root>
  <rux-option label="one"></rux-option>
  <rux-option label="two"></rux-option>
  <input class="aux-select aux-select-0" type="hidden">
</rux-select>
`;

exports[`rux-select renders option groups 1`] = `
<rux-select size="medium">
  <mock:shadow-root>
    <label aria-hidden="true" part="label">
      <span class="hidden">
        <slot name="label"></slot>
      </span>
    </label>
    <select class="rux-select rux-select--medium" part="select"></select>
    <div aria-hidden="true" class="hidden">
      <slot></slot>
    </div>
    <div class="hidden" part="error-text">
      <svg fill="none" height="14" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
        <path clip-rule="evenodd" d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z" fill="currentColor" fill-rule="evenodd"></path>
      </svg>
      <slot name="error-text"></slot>
    </div>
    <div class="hidden" part="help-text">
      <slot name="help-text"></slot>
    </div>
  </mock:shadow-root>
  <rux-option label="outside option"></rux-option>
  <rux-option-group label="Group one">
    <rux-option label="inside option"></rux-option>
  </rux-option-group>
  <rux-option label="outside option"></rux-option>
  <input class="aux-select aux-select-0" type="hidden">
</rux-select>
`;
